File: 1025-minimum-cost-for-tickets/minimum-cost-for-tickets.cpp
class Solution {
public:
    int noOfDays[3] = {1, 7, 30};
    unordered_map <int, int> memo;
    
    int dfs(int index, vector<int>& days, vector<int>& costs){

        if (memo.contains(index))
            return memo[index];
            
        memo[index] = INT_MAX;

        for (int i = 0; i < 3; i++){
            int j = index;
            while (j < days.size() && days[j] < days[index] +  noOfDays[i])
                j += 1;
            memo[index] = min(memo[index] , costs[i] + dfs(j, days, costs));
        }
        return memo[index] ;
    }

    int mincostTickets(vector<int>& days, vector<int>& costs) {
        memo[days.size()] = 0;
        return dfs(0, days, costs);
    }
};

File: 1025-minimum-cost-for-tickets/minimum-cost-for-tickets.py
class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        memo = {
            len(days): 0
        }

        for i in reversed(range(len(days))):
            if i in memo:
                return memo[i]

            memo[i] = float("inf")
            for cost, day in zip(costs, [1, 7, 30]):
                j = i
                while j < len(days) and days[j] < days[i] + day:
                    j += 1
                memo[i] =  min(memo[i], cost + memo[j])

        return memo[0]
        
        # top-down
        # memo = {
        #     len(days): 0
        # }

        # def dfs(i):
        #     if i in memo:
        #         return memo[i]

        #     memo[i] = float("inf")
        #     for cost, day in zip(costs, [1, 7, 30]):
        #         j = i
        #         while j < len(days) and days[j] < days[i] + day:
        #             j += 1
        #         memo[i] =  min(memo[i], cost + dfs(j))
        #     return memo[i]

        # return dfs(0)
            
                



        # result = float("inf")

        # def dfs(index, curr, p):
        #     nonlocal result
        #     if len(days) == index:
        #         result = min(result, sum(p))
        #         print(index, curr, p)
        #         return
        #     for cost in costs:
        #         curr += cost
        #         while index < len(days) and curr >= days[index]:
        #             index += 1
        #         dfs(index, days[index] - 1 if index < len(days) else curr, p + [cost])
        
        # dfs(0, 0, [])
        # return result